File: haskell/src/InputOutput.hs
module InputOutput
    ( demonstrateIO
    ) where

import System.IO (hFlush, stdout)

-- | Function to demonstrate input and output operations in Haskell
demonstrateIO :: IO ()
demonstrateIO = do
    putStrLn "=== Input and Output ==="
    
    -- Basic output
    putStrLn "Basic output using putStrLn"
    putStr "Basic output using putStr (no newline)"
    putStrLn ""
    
    -- Print function (shows the value's representation)
    print True
    print [1, 2, 3]
    print (1.5 :: Double)
    
    -- String concatenation in output
    let name = "Alice"
        age = 25
    putStrLn $ "Name: " ++ name ++ ", Age: " ++ show age
    
    -- Basic input (commented out for non-interactive demos)
    {-
    putStr "Enter your name: "
    hFlush stdout  -- Ensure prompt is shown before reading input
    inputName <- getLine
    putStrLn $ "Hello, " ++ inputName ++ "!"
    
    putStr "Enter your age: "
    hFlush stdout
    inputAge <- getLine
    putStrLn $ "In 10 years, you'll be " ++ show (read inputAge + 10) ++ " years old."
    -}
    
    -- File operations (demonstration with a temporary file)
    -- Write to file
    writeFile "test.txt" "Line 1\nLine 2\nLine 3\n"
    
    -- Read from file
    putStrLn "\nReading from file:"
    contents <- readFile "test.txt"
    putStr contents
    
    -- Read file line by line
    putStrLn "Reading file line by line:"
    lines' <- fmap lines (readFile "test.txt")
    mapM_ putStrLn lines'
    
    -- Append to file
    appendFile "test.txt" "Line 4 (appended)\n"
    
    -- Read updated file
    putStrLn "\nReading updated file:"
    newContents <- readFile "test.txt"
    putStr newContents
    
    -- Clean up
    writeFile "test.txt" ""  -- Clear file contents
    
    -- Do/Let notation
    putStrLn "\nDo/Let notation example:"
    let calculation = do
            let x = 5
                y = 7
            return $ x + y
    result <- return calculation
    putStrLn $ "5 + 7 = " ++ show result
    
    putStrLn "" -- Empty line

File: lua/src/input_output.lua
-- Module for demonstrating input and output operations in Lua
local M = {}

function M.demonstrate_io()
    print("=== Input and Output ===")
    
    -- Basic print output
    print("Basic print:", "Hello", "World", 123)
    
    -- String formatting with string.format
    local name = "Alice"
    local age = 25
    print(string.format("Formatted string: Name = %s, Age = %d", name, age))
    
    -- Writing to standard output with io.write
    io.write("This is written using io.write (no automatic newline)\n")
    
    -- String concatenation with ..
    print("String " .. "concatenation " .. "with .. operator")
    
    -- Reading input from console (commented out for non-interactive demos)
    --[[
    io.write("Enter your name: ")
    local user_name = io.read()
    print("Hello, " .. user_name)
    
    io.write("Enter your age: ")
    local user_age = tonumber(io.read())
    if user_age then
        print("In 10 years, you'll be " .. (user_age + 10))
    end
    --]]
    
    -- File operations
    -- Writing to a file
    local file = io.open("test.txt", "w")
    if file then
        file:write("Line 1\n")
        file:write("Line 2\n")
        file:write("Line 3\n")
        file:close()
    end
    
    -- Reading from a file
    print("\nReading from file:")
    file = io.open("test.txt", "r")
    if file then
        for line in file:lines() do
            print(line)
        end
        file:close()
    end
    
    -- Clean up
    os.remove("test.txt")
    
    print() -- Empty line
end

return M